File: packages/typia/src/programmers/internal/metadata_to_pattern.ts
import { Metadata } from "../../schemas/metadata/Metadata";

import { PatternUtil } from "../../utils/PatternUtil";

import { template_to_pattern } from "./template_to_pattern";

/** @internal */
export const metadata_to_pattern = (props: {
  top: boolean;
  metadata: Metadata;
}): string => {
  if (props.metadata.atomics.find((a) => a.type === "string") !== undefined)
    return "(.*)";

  const values: string[] = props.metadata.constants
    .map((c) => {
      if (c.type !== "string") return c.values.map((v) => v.toString());
      return (c.values.map((v) => v.value) as string[]).map((str) =>
        PatternUtil.escape(str),
      );
    })
    .flat();
  for (const a of props.metadata.atomics)
    if (a.type === "number" || a.type === "bigint")
      values.push(PatternUtil.NUMBER);
    else if (a.type === "boolean") values.push(PatternUtil.BOOLEAN);
  for (const { row } of props.metadata.templates)
    values.push(
      "(" +
        template_to_pattern({
          top: false,
          template: row,
        }) +
        ")",
    );

  const pattern: string =
    values.length === 1 ? values[0]! : "(" + values.join("|") + ")";
  return props.top ? PatternUtil.fix(pattern) : pattern;
};

File: packages/typia/native/core/programmers/iterate/metadata_to_pattern.go
package iterate

import (
	"fmt"
	"strings"

	nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
	nativeutils "github.com/samchon/typia/packages/typia/native/core/utils"
)

func metadata_to_pattern(props struct {
	top      bool
	metadata *nativemetadata.MetadataSchema
}) string {
	for _, atomic := range props.metadata.Atomics {
		if atomic.Type == "string" {
			return "(.*)"
		}
	}

	values := []string{}
	for _, constant := range props.metadata.Constants {
		if constant.Type != "string" {
			for _, value := range constant.Values {
				values = append(values, fmt.Sprint(value.Value))
			}
			continue
		}
		for _, value := range constant.Values {
			if str, ok := value.Value.(string); ok {
				values = append(values, nativeutils.PatternUtil.Escape(str))
			} else {
				values = append(values, nativeutils.PatternUtil.Escape(fmt.Sprint(value.Value)))
			}
		}
	}
	for _, atomic := range props.metadata.Atomics {
		if atomic.Type == "number" || atomic.Type == "bigint" {
			values = append(values, nativeutils.PatternUtil.NUMBER)
		} else if atomic.Type == "boolean" {
			values = append(values, nativeutils.PatternUtil.BOOLEAN)
		}
	}
	for _, template := range props.metadata.Templates {
		values = append(values, "("+template_to_pattern(struct {
			top      bool
			template []*nativemetadata.MetadataSchema
		}{
			top:      false,
			template: template.Row,
		})+")")
	}

	pattern := ""
	if len(values) == 1 {
		pattern = values[0]
	} else {
		pattern = "(" + strings.Join(values, "|") + ")"
	}
	if props.top {
		return nativeutils.PatternUtil.Fix(pattern)
	}
	return pattern
}
